### PR TITLE
Feature/take out the garbage

### PR DESCRIPTION
**Pull request by KevinDang74**
Please review my code.

### CODE REVIEW

File: challenge_take_out_the_garbage/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+## Uncomment this if the package has a setup.py. This macro ensures
+## modules and global scripts declared therein get installed
+## See http://ros.org/doc/api/catkin/html/user_guide/setup_dot_py.html
+catkin_python_setup()
+
+################################################
+## Declare ROS messages, services and actions ##
+################################################
+
+## To declare and build messages, services or actions from within this
```
Comment (LoyVanBeek) [inline challenge_take_out_the_garbage/CMakeLists.txt]: You can clean up all the outcommented stuff here

File: challenge_take_out_the_garbage/package.xml [hunk 2]
```diff
@@ -0,0 +1,65 @@
+<?xml version="1.0"?>
```
Comment (LoyVanBeek) [inline challenge_take_out_the_garbage/package.xml]: Same here, just remove all the outcommented stuff and please fill in the templated fields

File: challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+import rospy
+import smach
+
+# TU/e
+import robot_skills
+import robot_smach_states as states
+import robot_smach_states.util.designators as ds
+
+
+class DefaultGrabDesignator(ds.Designator):
```
Comment (LoyVanBeek) [inline challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out.py]: Is this general enough to be moved to robot_smach_states?

File: challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+            return None
+
+        # Sort the entities and return the closest one
+        base_pose = self._robot.base.get_location().frame
+        entities = sorted(entities, key=lambda e: e.distance_to_2d(base_pose.p))
+        rospy.loginfo(entities[0])
+        return entities[0]
+
+
+class GrabSingleItem(smach.StateMachine):
```
Comment (LoyVanBeek) [inline challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out.py]: Is this general enough to be moved to robot_smach_states?

File: challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out_the_garbage.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    """
+    def __init__(self, robot):
+        """ Initialization method
+
+        :param robot: robot api object
+        """
+        smach.StateMachine.__init__(self, outcomes=["succeeded", "failed", "aborted"])
+
+        # Create designators
+        trashbin_id = "trashbin"
```
Comment (LoyVanBeek) [inline challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out_the_garbage.py]: I'd put these variables in the knowledge, same as the drop zone.

File: challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        :param robot: robot object
+        :param surface_designator: designator for the object to grab from
+        :param area_description: string with id of the area where the object should be located in
+        """
+        super(DefaultGrabDesignator, self).__init__(resolve_type=robot_skills.util.entity.Entity)
+
+        self._robot = robot
+        self._surface_designator = surface_designator
+        self.area_description = area_description
+        self._debug = debug
```
Comment (MatthijsBurgh) [inline challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out.py]: please remove the debug arguments

File: challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out_the_garbage.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
+# ROS
+import smach
+
+# Robot smach states
+import robot_smach_states as states
+from robocup_knowledge import load_knowledge
+import robot_smach_states.util.designators as ds
+from challenge_take_out_the_garbage.take_out import DefaultGrabDesignator, GrabSingleItem, TakeOut
```
Comment (MatthijsBurgh) [inline challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out_the_garbage.py]: Absolute imports 👍

File: challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out_the_garbage.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        smach.StateMachine.__init__(self, outcomes=["succeeded", "failed", "aborted"])
+
+        # Create designators
+        trashbin_id = "trashbin"
+        trashbin_designator = ds.EdEntityDesignator(robot=robot,
+                                                    id=trashbin_id)
+        trash_designator = DefaultGrabDesignator(robot=robot,
+                                                 surface_designator=trashbin_designator,
+                                                 area_description="on_top_of",
+                                                 debug=True)
```
Comment (MatthijsBurgh) [inline challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out_the_garbage.py]: Remove the debug argument here too

File: challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out_the_garbage.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        trashbin_id2 = "trashbin2"
+        trashbin_designator2 = ds.EdEntityDesignator(robot=robot,
+                                                     id=trashbin_id2)
+        trash_designator2 = DefaultGrabDesignator(robot=robot,
+                                                  surface_designator=trashbin_designator2,
+                                                  area_description="on_top_of",
+                                                  debug=True)
```
Comment (MatthijsBurgh) [inline challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out_the_garbage.py]: Remove the debug argument here too

File: challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+
+        self._robot = robot
+        self._place_designator = place_designator
+
+    def execute(self, userdata=None):
+
+        # ToDo: turn this into a proper statemachine
+        arm = None
+        # See if there's an arm holding something
+        for k, v in self._robot.arms.iteritems():
```
Comment (MatthijsBurgh) [inline challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out.py]: I think this should be different nowadays, check with @alberth

File: challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+        :param area_description: string with id of the area where the object should be located in
+        """
+        super(DefaultGrabDesignator, self).__init__(resolve_type=robot_skills.util.entity.Entity)
+
+        self._robot = robot
+        self._surface_designator = surface_designator
+        self.area_description = area_description
+        self._debug = debug
+
+    def resolve(self):
```
Comment (MatthijsBurgh) [inline challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out.py]: I think the `resolve()` shouldn't be overwritten, but the `_resolve()` @LoyVanBeek am I right?

(I think you copied this from the storing groceries, so we would need to check all designators around

File: challenge_take_out_the_garbage/scripts/challenge_take_out_the_garbage [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+# TU/e Robotics
+from robot_skills.util.robot_constructor import robot_constructor
+
+# Challenge example
+from challenge_take_out_the_garbage.take_out_the_garbage import TakeOutGarbage
+
+if __name__ == "__main__":
+
+    rospy.init_node("state_machine_garbage")
+    robot = robot_constructor("amigo")   # ToDo: Make configurable
```
Comment (MatthijsBurgh) [inline challenge_take_out_the_garbage/scripts/challenge_take_out_the_garbage]: Use https://github.com/tue-robotics/tue_robocup/blob/master/robot_smach_states/src/robot_smach_states/util/startup.py like all challenges

File: challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out.py (hunk 6 above)
Comment (LoyVanBeek) [reply challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out.py]: Why?

File: challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out.py (hunk 11 above)
Comment (LoyVanBeek) [reply challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out.py]: Correct, subclasses should implement `_resolve`

File: challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out.py (hunk 10 above)
Comment (alberth) [reply challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out.py]: Indeed, robot has a `get_arm` where you can ask for an arm holding anything.

File: challenge_take_out_the_garbage/package.xml [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+  <!-- Use buildtool_depend for build tool packages: -->
+  <!--   <buildtool_depend>catkin</buildtool_depend> -->
+  <!-- Use exec_depend for packages you need at runtime: -->
+  <!--   <exec_depend>message_runtime</exec_depend> -->
+  <!-- Use test_depend for packages you need only for testing: -->
+  <!--   <test_depend>gtest</test_depend> -->
+  <!-- Use doc_depend for packages you need only for building documentation: -->
+  <!--   <doc_depend>doxygen</doc_depend> -->
+  <buildtool_depend>catkin</buildtool_depend>
+  <build_depend>robot_skills</build_depend>
```
Comment (MatthijsBurgh) [inline challenge_take_out_the_garbage/package.xml]: build_depend+build_export_depend+exec_depend=depend

File: challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out.py (hunk 6 above)
Comment (MatthijsBurgh) [reply challenge_take_out_the_garbage/src/challenge_take_out_the_garbage/take_out.py]: Because if self.debug is commented out on line 34.

**Comment by ar13pit** [comment]
@MatthijsBurgh @LoyVanBeek @SamAlexandrov is this PR still required?

**Comment by LoyVanBeek** [comment]
This PR is still required, the commits are not yet on `master` nor `rwc2019`

**Comment by ar13pit** [comment]
These commits are already in `master` through #826

**Pull request closed without merging**